# TECH 2 mandatory assignment - Part B

Write your solution for Part B in this notebook.

In [2]:
import pandas as pd

 # read_csv function loads the csv file with data, and we store it in df (dataframe
# iloc retrieves data from a certain location, y coordinate for columns
# The two first columns have boxes with no values as the length is 10 000, dropna ensures that Nan (not a number) values are not included
# tolist() stores the data in a list


df = pd.read_csv('Data.csv')                      
column_100 = df.iloc[:, 0].dropna().tolist()       
column_1000 = df.iloc[:, 1].dropna().tolist()
column_10000 = df.iloc[:, 2].tolist()


In [3]:
 # Import loops function from Part A, and use the function for the 3 datasets
from part_A import std_loops
print(std_loops(column_100))
print(std_loops(column_1000))
print(std_loops(column_10000))

1.4142135623730951
0.2830566925790975
0.2847606502984571
0.2854126090040637


In [4]:
 # Import function with built in functions, and use the function for the 3 datsets
from part_A import std_builtin
print(std_builtin(column_100))
print(std_builtin(column_1000))
print(std_builtin(column_10000))

0.2830566925790975
0.2847606502984576
0.28541260900406346


In [5]:
 # Import numpy, and use numpy's standard deviation function for the 3 datasets
import numpy as np
print(np.std(column_100))
print(np.std(column_1000))
print(np.std(column_10000))

0.2830566925790975
0.28476065029845754
0.28541260900406346


Use timeit tool to measure the time used by the 3 functions for the 3 datasets

In [7]:
%%timeit
std_loops(column_100)

8.1 μs ± 34.4 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [8]:
%%timeit 
std_loops(column_1000)

87.4 μs ± 301 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [9]:
%%timeit
std_loops(column_10000)

888 μs ± 3.79 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [10]:
%%timeit
std_builtin(column_100)

6.01 μs ± 4.25 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [11]:
%%timeit
std_builtin(column_1000)

61.7 μs ± 219 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [12]:
%%timeit
std_builtin(column_10000)

613 μs ± 3.87 μs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


In [13]:
%%timeit
np.std(column_100)

10.2 μs ± 96.9 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)


In [14]:
%%timeit
np.std(column_1000)

30.6 μs ± 52.4 ns per loop (mean ± std. dev. of 7 runs, 10,000 loops each)


In [15]:
%%timeit
np.std(column_10000)

230 μs ± 435 ns per loop (mean ± std. dev. of 7 runs, 1,000 loops each)


The output of timeit will vary slightly. The output I recieved when using timeit was as follows:
Std_loops for 100 numbers: 8.1 μs ± 34.4 ns per loop 
Std_loops for 1000 numbers: 87.4 μs ± 301 ns per loop
Std_loops for 10000 numbers: 888 μs ± 3.79 μs per loop

Std_builtin for 100 numbers: 6.01 μs ± 4.25 ns per loop
Std_builtin for 1000 numbers: 61.7 μs ± 219 ns per loop
Std_builtin for 10000 numbers: 613 μs ± 3.87 μs per loop

np.std for 100 numbers: 10.2 μs ± 96.9 ns per loop
np.std for 1000 numbers: 30.6 μs ± 52.4 ns per loop
np.std for 10000 numbers: 230 μs ± 435 ns per loop

The function using built in functions was considerably faster than the function using loops for all 3 datasets. This concludes that built in functions are more efficient than loops. The numpy solution is slower than both of the functions for the first dataset with 100 numbers, but it becomes much faster than the functions for the second and third dataset. Numpy handles 10 times the amount of data from dataset 1 to 2 but uses approximately three times the amount of time. Numpy handles 10 times the amount of data from data set 2 to 3 but uses approximately 8 times the amount of time. This suggests that the optimization of the numpy solution kicks in when handling larger datasets, but is out-competed by the simplicity of the self-made functions. 

In conclusion, the function with built in functions is fastest for lists with a couple hundred numbers and less, and the numpy function is fastest for larger datasets.